[API] Feature/job failure policy

clamoriniere1A opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**: Implements the Backoff policy and failed pod limit defined in https://github.com/kubernetes/community/pull/583

**Which issue this PR fixes**: 
fixes #27997, fixes #30243

**Special notes for your reviewer**:
This is a WIP PR, I updated the api batchv1.JobSpec in order to prepare the backoff policy implementation in the JobController.

**Release note**:
```release-note
Add backoff policy and failed pod limit for a job
```

kargakis commented:
@kubernetes/sig-apps-api-reviews

clamoriniere1A commented:
CLA signed

spiffxp commented:
/ok-to-test

clamoriniere1A commented:
Hi @soltysh 

This PR is linked to the proposal “Backoff policy and failed pod limit”: https://github.com/kubernetes/community/pull/583

I started to update the jobcontroller.syncJob() logic, and I have 2 questions regarding the attended behaviours:
**1) Do we need to delete all the active running Pod in case the ```Backofflimit``` have been reached?** like it is done when we cross the “Job Deadine”.
(code link: https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/controller/job/jobcontroller.go#L521 )

**2) In order to handle properly the deletion of failed pods when the ```Job.Spec. FailedPodsLimit``` is reached.** 
I needed to add a new counter ```FailedAndDeleted``` in the ```Job.Status struct``` ( https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/apis/batch/types.go#L179 ). The new counter allows me to calculate the total number of failed Pod: “Failed and Active” + “Failed and deleted”.
**Do you think is a good approach?**

clamoriniere1A commented:
/test pull-kubernetes-e2e-gce-etcd3

clamoriniere1A committed:
update API v1 Job object

Add new fields in api v1.JobSpec object for backoff policy
- BackoffLimit
- FailedPodsLimit

fixes: https://github.com/kubernetes/community/pull/583

soltysh commented:
/assign

soltysh commented:
@clamoriniere1A

> 1) Do we need to delete all the active running Pods in case the Backofflimit have been reached? like it is done when the Job cross the “Job Deadine”.

Yes. We need to stop whatever is already running. 

> 2) In order to handle properly the deletion of failed pods when the Job.Spec. FailedPodsLimit is reached.
> I needed to add a new counter FailedAndDeleted in the Job.Status struct ( https://github.com/clamoriniere1A/kubernetes/blob/feature/job_failure_policy/pkg/apis/batch/types.go#L179 ). The new counter allows me to calculate the total number of failed Pod: “Failed and Active” + “Failed and deleted”.
> Do you think it is a good approach?

I don't think it's needed. The `Job.Status.Failed` value should have all the pods that failed during the overall run. Not just currently available.

soltysh commented on pkg/apis/batch/v1/defaults_test.go:
```diff
@@ -115,47 +123,95 @@ func TestSetDefaultJob(t *testing.T) {
+					Completions:     newInt32(1),
+					Parallelism:     newInt32(1),
+					BackoffLimit:    newInt32(2),
+					FailedPodsLimit: newInt32(1),
```
That seems wrong, if you set it explicitly to 2, why you're expecting 1, here?

clamoriniere1A replied:
Right, 
The fact is that I mess up some commits split. I did include in this commit the test logic update that is checking the value: BackoffLimit and FailedPodsLimit.

soltysh commented on pkg/apis/batch/types.go:
```diff
@@ -160,6 +173,10 @@ type JobStatus struct {
 	// +optional
 	Failed int32
+
+	// The number of pods which reached phase Failed and deleted.
```
Not needed, the above `Failed` field should hold information about all the failed pods. Currently it's the job controller that is broken, see https://github.com/kubernetes/kubernetes/issues/28486 for details.

clamoriniere1A replied:
Ok I will remove this field

clamoriniere1A commented:
/test pull-kubernetes-e2e-kops-aws

clamoriniere1A commented:
/test pull-kubernetes-e2e-kops-aws

clamoriniere1A commented:
/test pull-kubernetes-verify

clamoriniere1A commented:
/test pull-kubernetes-unit

soltysh commented:
/lgtm
/approve

soltysh commented:
@erictune needs your sign-off for the API changes

kargakis commented:
/test pull-kubernetes-bazel-build 
/test pull-kubernetes-bazel-test

smarterclayton commented:
Based on reading of the proposal and conventions this is API approved.  The PR that enables failedpodslimit will still need API review, although I confirmed it matches the proposal.

/approve

clamoriniere1A committed:
Generate files from v1.JobSpec modification

This commit contains the new version of generated api files linked
to the v1.JobSpec modifications in the previous commit after
"make update"

soltysh commented:
/lgtm

soltysh commented:
/retest

clamoriniere1A commented:
/test pull-kubernetes-kubemark-e2e-gce

soltysh commented:
/test pull-kubernetes-kubemark-e2e-gce

soltysh commented:
/test pull-kubernetes-kubemark-e2e-gce

soltysh commented:
/retest

CaoShuFeng commented:
/test pull-kubernetes-verify

kargakis commented:
Is there a PR with the controller changes?

cedriclam commented:
@kargakis, yes #51153 contains controller changes

kargakis commented:
merci
